### ISSUE TITLE
Statements about unicornafl

### ISSUE THREAD

**Issue by wtdcode** [icicle-emu/icicle-emu#29]
Hello guys! Nice work! I have a few questions while reading your [paper}(https://arxiv.org/pdf/2301.13346.pdf).

I notice this paragraph:

```
Notably, in AFL++’s Unicorn mode, branch coverage instrumentation is not implemented
using Unicorn’s hooking API, since the instrumentation is highly
performance sensitive and the hooking API imposes additional
overheads.
```

Notably unicornafl had already used Unicorn standard hook API to instrument the target binary and as far as I benchmarked, the performance has no obvious difference. Besides, IIRC, fuzzware is based on uncorn1, while Unicorn2 has an average of 15% performance boost which also benefits unicornafl. Which Unicorn version did you use in this paper?

**Comment by mchesser** [comment]
Hi, thanks for the interest in the paper.

I think I originally wrote that statement back in 2021 when this was the case (and I think Unicorn 2 had yet to be released). I believe the original version of AFL++ I used was [AFLplusplus@23f7bee](https://github.com/AFLplusplus/AFLplusplus/tree/23f7bee81c46ad4f0f65fa56d08064ab5f1e2e6f), which would have corresponded to me reading the code here: [unicornafl/qemu/target-i386/translate.c:8759](https://github.com/AFLplusplus/unicornafl/blob/fb2fc9f25df32f17f6b6b859e4dbd70f9a857e0c/qemu/target-i386/translate.c#L8759-L8762)

The paper was updated to discuss Unicorn 2 when it was released, and I did rerun benchmarks vs QEMU with a newer version of AFL++. However, unfortunately I guess I missed the fact that AFL++ had upgraded its Unicorn mode in the meantime.

In Icicle, I did observe significant performance differences with inline instrumentation vs hook based instrumentation, e.g. I quick experiment I did now running `/bin/date` in Icicle yields:

* No instrumentation: ~5778.75 execs/sec
* Inline hit-counts: ~5039.14 execs/sec
* Trampoline based hit-counts: ~3988.27 execs/sec

(Obviously this is a micro-benchmark, this is running without any fuzzer involved).

These performance results, in combination with the fact that the original version did not make use of the hooking API, together with some papers I read at the time (e.g. [zAFL](https://www.usenix.org/system/files/sec21-nagy.pdf)), where why that statement was included in the paper.

I still believe that branch coverage is highly performance sensitive -- however, maybe the fact that the original implementation non-standard implementation _still_ involves calling a trampoline, means is less of a difference. I think it would be worth testing whether using implementation that directly generates TcgOps would improve performance in unicornafl.


The paper only really directly compares with Unicorn in the comparisons against Fuzzware -- which like you have noted is Unicorn1.

**Comment by wtdcode** [comment]
Thanks for your explanation! It’s really a long journey ;)

Regarding your benchmark, while unicornafl moves to hook based API, we also implement the inline optimizations in Unicorn2, which is the key factor of not losing speed. This also proves that hooking itself is not the main reason for overhead.

**Comment by mchesser** [comment]
Yeah, Unicorn has certainly improved a lot recently. There are always endless optimizations that can be done to emulators, working out the ones that have a meaningful impact is always challenging. I found working directly with the lifted code to work really nicely and efficiently, but there are always other ways of doing things.

Do you have a link to the code/commit/pr implementing inline optimizations in Unicorn2? I'd be interested to see the implementation.

**Comment by wtdcode** [comment]
https://github.com/unicorn-engine/unicorn/commit/87a391d549a339b5d8109c10c6b16bd95130a923

This is the start of optimizations for unicornafl which inlines the instrumentation.

**Comment by mchesser** [comment]
Thanks. Closing this issue for now, but feel free to ask any other questions.

I may eventually end up linking this discussion from the paper repo so it is clear what the latest version of Unicorn supports.
